File: lab4/lab4.py
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

def invertTree(root):
    if root:
        # Swap left and right subtrees
        root.left, root.right = root.right, root.left
        # Recursively invert left and right subtrees
        invertTree(root.left)
        invertTree(root.right)
    return root

# Create a sample binary tree
def create_sample_tree():
    root = TreeNode(1)
    root.left = TreeNode(2)
    root.right = TreeNode(3)
    root.left.left = TreeNode(4)
    root.left.right = TreeNode(5)
    return root

# Invert the tree
sample_tree = create_sample_tree()
inverted_tree = invertTree(sample_tree)

# Print the inverted tree (should be 1 -> 3 -> 2 -> 5 -> 4)
def print_tree(root):
    if root:
        print(root.val)
        print_tree(root.left)
        print_tree(root.right)

print_tree(inverted_tree)

File: lab4/lab4.java
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int val) {
        this.val = val;
    }
}

public class TreeInverter {
    public TreeNode invertTree(TreeNode root) {
        if (root != null) {
            // Swap left and right subtrees
            TreeNode temp = root.left;
            root.left = root.right;
            root.right = temp;

            // Recursively invert left and right subtrees
            invertTree(root.left);
            invertTree(root.right);
        }
        return root;
    }

    // Example usage
    public static void main(String[] args) {
        // Create a sample binary tree
        //       1
        //      / \
        //     2   8
        //    / \  /\
        //   3   4  5 6
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(2);
        root.right = new TreeNode(8);
        root.left.left = new TreeNode(3;
        root.left.right = new TreeNode(4);
        root.left.left = new TreeNode(5);
        root.left.right = new TreeNode(6);
        // Invert the tree
        TreeInverter inverter = new TreeInverter();
        TreeNode invertedRoot = inverter.invertTree(root);

        // Print the inverted tree (should be 1 -> 3 -> 2 -> 5 -> 4)
        printTree(invertedRoot);
    }

    private static void printTree(TreeNode root) {
        if (root != null) {
            System.out.println(root.val);
            printTree(root.left);
            printTree(root.right);
        }
    }
}
